src/thread.rs

```diff
@@ -120,22 +196,51 @@ where
     // Publish the result and unblock the waiter. We need to do this now, because once this
     // closure completes, the Execution will consider this task Finished and invoke the
     // scheduler.
-    *result.lock().unwrap() = Some(Ok(ret));
+    *result.borrow_mut() = Some(Ok(ret));
     ExecutionState::with(|state| {
         if let Some(waiter) = state.current_mut().take_waiter() {
             state.get_mut(waiter).unblock();
         }
     });
 }
 
+/// An owned permission to join on a scoped thread (block on its termination).
+///
+/// See [`Scope::spawn`] for details.
+#[derive(Debug)]
+pub struct ScopedJoinHandle<'scope, T> {
+    handle: JoinHandle<T>,
+    _marker: PhantomData<&'scope T>,
+}
+
+#[derive(Clone, Debug)]
+struct ScopeData {
+    num_running_threads: Rc<RefCell<usize>>,
+    main_task: TaskId,
+}
```

Comment:
I don't see the need for `ScopeData` at all, why cannot the fields just be inlined into `Scope`?